go/adbc/driver/snowflake/bulk_ingestion.go

```diff
@@ -0,0 +1,546 @@
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(concurrency)
+
+	readyCh := make(chan struct{}, 1)
+	stopCh := make(chan interface{})
+
+	// Handler to signal that new data has been uploaded.
+	// Executing COPY will be a no-op if this has not been called since the last COPY was dispatched, so we wait for the signal.
+	readyFn := func() {
+
+		// readyFn is a no-op if the shutdown signal has already been recieved
+		select {
+		case _, ok := <-stopCh:
+			if !ok {
+				return
+			}
+		default:
+			// Continue
+		}
+
+		// readyFn is a no-op if it already knows that at least 1 file is ready to be loaded
+		select {
+		case readyCh <- struct{}{}:
+		default:
+			return
+		}
+	}
+
+	// Handler to signal that ingestion pipeline has completed successfully.
+	// Executes COPY to finalize ingestion (may no-op if all files already loaded by bg workers) and waits for all COPYs to complete.
+	stopFn := func() error {
+		defer cancel()
+		close(stopCh)
+		close(readyCh)
+
+		g.Go(func() error {
+			_, err := cn.ExecContext(ctx, copyQuery, []driver.NamedValue{{Value: tableName}})
+			return err
+		})
+
+		return g.Wait()
+	}
+
+	// Handler to signal that ingestion pipeline failed and COPY operations should not proceed.
+	// Stops the dispatch of new bg workers and cancels all in-progress COPY operations.
+	cancelFn := func() {
+		defer cancel()
+		close(stopCh)
+		close(readyCh)
+	}
+
+	go func() {
+		for {
+
+			// Check for shutdown signals before starting a new COPY
+			select {
+			case <-stopCh:
+				return
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			g.Go(func() error {
+				// Block until there is at least 1 new file available for copy, or it's time to shutdown
+				select {
+				case _, ok := <-readyCh:
+					if !ok {
+						return nil
+					}
+					// Proceed to start a new COPY job
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+
+				_, err := cn.ExecContext(ctx, copyQuery, []driver.NamedValue{{Value: tableName}})
+
+				return err
+			})
+		}
+	}()
+
+	return readyFn, stopFn, cancelFn
+}
+
+func countRowsInTable(ctx context.Context, db *sql.DB, tableName string) (int64, error) {
+	var nrows int64
+
+	row := db.QueryRowContext(ctx, countQuery, tableName)
+	if err := row.Scan(&nrows); err != nil {
+		return 0, err
+	}
+
+	return nrows, nil
+}
+
+// Initializes a sync.Pool of *bytes.Buffer.
+// Extra space is preallocated so that the Parquet footer can be written after reaching target file size without growing the buffer
+func newBufferPool(size int) (getFn func() *bytes.Buffer, putFn func(*bytes.Buffer)) {
+	buffers := sync.Pool{
+		New: func() interface{} {
+			extraSpace := 1 * megabyte // TODO(joellubi): Generally works, but can this be smarter?
+			buf := make([]byte, 0, size+extraSpace)
+			return bytes.NewBuffer(buf)
+		},
+	}
+
+	getFn = func() *bytes.Buffer {
+		return buffers.Get().(*bytes.Buffer)
+	}
+
+	putFn = func(buf *bytes.Buffer) {
+		buf.Reset()
+		buffers.Put(buf)
+	}
+
+	return getFn, putFn
+}
```

Comment:
i'd be curious if it wouldn't be more performant to have an object with a `sync.Pool` member who has get and put functions which use these impls rather than using two separate closures? The closures likely prevent inlining as opposed to it being an object

Reply:
Sure, I just pushed this change. Is that what you had in mind?

Reply:
yup pretty much. just more convenient than having to pass around both functions